src/Features/Protocol/LanguageServices/RoslynLanguageService.cs

```diff
@@ -0,0 +1,673 @@
+                    return null;
+                }
+
+                var symbols = symbolItem.NavigationSymbolId.Resolve(compilation, cancellationToken: cancellationToken);
+                var symbol = symbols.Symbol;
+
+                if (symbol == null)
+                {
+                    if (symbolItem.NavigationSymbolIndex < symbols.CandidateSymbols.Length)
+                    {
+                        symbol = symbols.CandidateSymbols[symbolItem.NavigationSymbolIndex.Value];
+                    }
+                    else
+                    {
+                        return null;
+                    }
+                }
+
+                var location = symbol.Locations.FirstOrDefault(l => l.SourceTree.Equals(tree));
+                return location ?? symbol.Locations.FirstOrDefault();
+            }
+
+            async Task<List<DocumentSymbol>> GetChildrenAsync(IEnumerable<NavigationBarItem> items)
+            {
+                var list = new List<DocumentSymbol>();
+                foreach (var item in items)
+                {
+                    list.Add(await CreateDocumentSymbolAsync(item).ConfigureAwait(false));
+                }
+
+                return list;
+            }
+
+            async Task<ISymbol> GetSymbolAsync(Location location)
+            {
+                var model = compilation.GetSemanticModel(location.SourceTree);
+                var root = await model.SyntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+                var node = root.FindNode(location.SourceSpan);
+
+                while (node != null)
+                {
+                    var symbol = model.GetDeclaredSymbol(node);
+                    if (symbol != null)
+                    {
+                        return symbol;
+                    }
+
+                    node = node.Parent;
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Answers a workspace symbols request by providing a list of symbols found in a given workspace.
+        /// https://microsoft.github.io/language-server-protocol/specification#workspace_symbol
+        /// </summary>
+        /// <param name="solution">the current solution.</param>
+        /// <param name="request">the workspace request with the query to invoke.</param>
+        /// <param name="cancellationToken">a cancellation token.</param>
+        /// <returns>a list of symbols in the workspace.</returns>
+        public async Task<SymbolInformation[]> GetWorkspaceSymbolsAsync(Solution solution, WorkspaceSymbolParams request, CancellationToken cancellationToken)
+        {
+            var symbols = ArrayBuilder<SymbolInformation>.GetInstance();
+
+            if (solution == null)
+            {
+                return symbols.ToArrayAndFree();
+            }
+
+            var searchTasks = solution.Projects.Select(
+                p => Task.Run(() => SearchProjectAsync(p), cancellationToken)).ToArray();
```

Comment:
1. this is not desirable if this is inside VS.  this will likely churn a ton of memory in parallel, and can cause slowdowns.  if this is just in a server process, that's likely ok.
2. FindRefernces searches by cliques to better ensure the same objects (like compilations) are not recomputed over and over again.  that may be necessary here.
3. it's very weird to not see cancellatoinToken passed to SearchProjectAsync.  I assume that's because you're capturing... but that makes the code very hard to validate from this location.

Reply:
we are taking live share's implementation. and it was like this for a while and live share has been used by people for more than a year so we will keep existing implementation as they are.